lib/pool-management/constructs/test-result-workflow.ts

```diff
@@ -0,0 +1,147 @@
+  public stateMachine: sfn.StateMachine;
+
+  constructor(scope: cdk.Construct, id: string, props: TestResultWorkflowProps) {
+    super(scope, id);
+
+    const forwardTestResultFunction = new lambdaNodejs.NodejsFunction(this, "forward-test-result-function", {
+      entry: path.join(__dirname, "../lambdas", "forward-test-results.ts"),
+      handler: "handler",
+      bundling: { externalModules: [], sourceMap: true, minify: true },
+      runtime: lambda.Runtime.NODEJS_14_X,
+      tracing: lambda.Tracing.ACTIVE,
+      logRetention: logs.RetentionDays.ONE_DAY,
+      timeout: cdk.Duration.seconds(12),
+      initialPolicy: [
+        new iam.PolicyStatement({
+          actions: ["mobiletargeting:CreateCampaign"],
+          resources: [`${props.pinpointApplication.attrArn}`, `${props.pinpointApplication.attrArn}/segments/*`],
+        }),
+      ],
+    });
+
+    forwardTestResultFunction.addEnvironment("POOL_TABLE_NAME", props.poolTable.tableName);
+    forwardTestResultFunction.addEnvironment("PINPOINT_APPLICATION_ID", props.pinpointApplication.ref);
+    props.poolTable.grantReadData(forwardTestResultFunction);
+
+    const campaignStatusFunction = new lambdaNodejs.NodejsFunction(this, "campaign-status-function", {
+      entry: path.join(__dirname, "../lambdas", "campaign-status.ts"),
+      handler: "handler",
+      bundling: { externalModules: [], sourceMap: true, minify: true },
+      runtime: lambda.Runtime.NODEJS_14_X,
+      tracing: lambda.Tracing.ACTIVE,
+      logRetention: logs.RetentionDays.ONE_DAY,
+      timeout: cdk.Duration.seconds(12),
+      initialPolicy: [
+        new iam.PolicyStatement({
+          actions: ["mobiletargeting:GetCampaignActivities"],
+          resources: [`${props.pinpointApplication.attrArn}`, `${props.pinpointApplication.attrArn}/campaigns/*`],
+        }),
+      ],
+    });
+    campaignStatusFunction.addEnvironment("PINPOINT_APPLICATION_ID", props.pinpointApplication.ref);
+
+    const forwardTestResultTask = new tasks.LambdaInvoke(this, "forward-test-result", {
+      lambdaFunction: forwardTestResultFunction,
+      payload: sfn.TaskInput.fromObject({
+        "tenant.$": "$.tenant",
+        "poolName.$": "$.poolName",
+        "testResult.$": "$.testResult",
+      }),
+      outputPath: "$.Payload",
+    });
+
+    const checkCampaignStatus = new tasks.LambdaInvoke(this, "check-campaign-status", {
+      lambdaFunction: campaignStatusFunction,
+      payload: sfn.TaskInput.fromObject({
+        "campaignId.$": "$.campaignId",
+      }),
+      outputPath: "$.Payload",
+    });
+
+    const waitForCampaignState = new sfn.Wait(this, "wait-for-campaign", {
+      time: sfn.WaitTime.duration(cdk.Duration.seconds(10)),
+    });
+
+    const logGroup = new logs.LogGroup(this, "test-result-processing-log-group", {
+      retention: logs.RetentionDays.ONE_DAY,
+    });
+
+    const stateMachineDefinition = forwardTestResultTask.next(
+      waitForCampaignState.next(
```

Comment:
Is the campaign creation an asynchronous process? Just asking since it's already "awaited" on in the lambda.

Reply:
Yes it is. It takes some time until a campaign is "successful" meaning that SMS are send out. This is based on a reference implementation from AWS: https://github.com/aws-samples/digital-user-engagement-reference-architectures#Automatic-Amazon-Pinpoint-Campaign-Creation